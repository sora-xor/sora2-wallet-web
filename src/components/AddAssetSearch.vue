<template>
  <div class="asset-search">
    <s-input
      class="asset-search-input"
      :maxlength="100"
      :placeholder="t(`addAsset.${AddAssetTabs.Search}.placeholder`)"
      border-radius="mini"
      v-model="search"
      @input="handleSearch"
    />
    <div class="asset-search-list" v-if="!!foundAssets">
      <div class="asset-search-list_info" v-if="!(search || foundAssets.length)">
        {{ t(`addAsset.${AddAssetTabs.Search}.info`) }}
      </div>
      <div v-if="search && !foundAssets.length" class="asset-search-list_empty">
        {{ t(`addAsset.${alreadyAttached ? 'alreadyAttached' : 'empty'}`) }}
      </div>
      <div
        class="asset s-flex"
        v-for="asset in foundAssets"
        :key="asset.address"
        :class="{ 'selected': (selectedAsset || {}).address === asset.address }"
        @click="handleSelectAsset(asset)"
      >
        <i :class="getAssetClasses(asset.address)" />
        <div class="asset-description s-flex">
          <div class="asset-description_symbol">{{ asset.symbol }}</div>
          <div class="asset-description_info">{{ formatName(asset) }}
            <s-tooltip :content="t('assets.copy')">
              <span class="asset-id" @click="handleCopy(asset)">({{ getFormattedAddress(asset) }})</span>
            </s-tooltip>
          </div>
        </div>
      </div>
    </div>
    <div class="asset-search-actions s-flex">
      <s-button @click="handleBack">
        {{ t('addAsset.cancel') }}
      </s-button>
      <s-button type="primary" :disabled="!selectedAsset" @click="handleAddAsset">
        {{ t('addAsset.action') }}
      </s-button>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator'
import { Action, Getter } from 'vuex-class'
import { AccountAsset, Asset, KnownAssets, KnownSymbols } from '@sora-substrate/util'

import TranslationMixin from './mixins/TranslationMixin'
import { AddAssetTabs, RouteNames } from '../consts'
import { copyToClipboard, formatAddress, getAssetIconClasses } from '../util'
import { NamedAsset } from '../types'

@Component
export default class AddAssetSearch extends Mixins(TranslationMixin) {
  readonly AddAssetTabs = AddAssetTabs
  readonly KnownSymbols = KnownSymbols

  @Getter assets!: Array<Asset>
  @Getter accountAssets!: Array<AccountAsset>
  @Action navigate
  @Action getAssets
  @Action addAsset

  search = ''
  selectedAsset: NamedAsset | null = null
  foundAssets: Array<NamedAsset> = []
  alreadyAttached = false

  mounted (): void {
    this.getAssets().then(this.handleSearch)
  }

  handleSearch (value?: string): void {
    this.alreadyAttached = false
    const assets = this.assets
      .filter(asset => !this.accountAssets.find(accountAsset => accountAsset.address === asset.address))
      .map(asset => {
        const knownAsset = KnownAssets.get(asset.address)
        return { ...asset, name: knownAsset ? this.t(`assetNames.${asset.symbol}`) : '' }
      })
    if (!value || !value.trim()) {
      this.foundAssets = assets
      this.selectedAsset = null
      return
    }
    const search = value.trim().toLowerCase()
    const attached = this.accountAssets.find(({ symbol, address }) => {
      const knownAsset = KnownAssets.get(address)
      return address.toLowerCase() === search || (symbol || '').toLowerCase() === search || (knownAsset && this.t(`assetNames.${symbol}`).toLowerCase() === search)
    })
    if (attached) {
      this.alreadyAttached = true
      this.foundAssets = []
      this.selectedAsset = null
      return
    }
    this.foundAssets = assets.filter(({ name, symbol, address }) => {
      return address.toLowerCase() === search || symbol.toLowerCase().includes(search) || name.toLowerCase().includes(search)
    })
    if (this.selectedAsset && !this.foundAssets.find(({ address }) => (this.selectedAsset || {}).address === address)) {
      this.selectedAsset = null
    }
  }

  formatName (asset: NamedAsset): string {
    if (!asset.name || asset.name.includes('assetNames')) {
      return asset.symbol
    }
    return asset.name
  }

  handleSelectAsset (asset: NamedAsset): void {
    this.selectedAsset = asset
  }

  handleBack (): void {
    this.navigate({ name: RouteNames.Wallet })
  }

  async handleAddAsset (): Promise<void> {
    await this.addAsset({ address: (this.selectedAsset || {}).address })
    this.navigate({ name: RouteNames.Wallet })
    this.$emit('add-asset')
  }

  getAssetClasses (address: string): string {
    return getAssetIconClasses(address)
  }

  getFormattedAddress (asset: NamedAsset): string {
    return formatAddress(asset.address, 10)
  }

  async handleCopy (asset: NamedAsset): Promise<void> {
    try {
      await copyToClipboard(asset.address)
      this.$notify({
        message: this.t('assets.successCopy', { symbol: asset.symbol }),
        type: 'success',
        title: ''
      })
    } catch (error) {
      this.$notify({
        message: `${this.t('warningText')} ${error}`,
        type: 'warning',
        title: ''
      })
    }
  }
}
</script>

<style scoped lang="scss">
@import '../styles/icons';

$asset-item-height: 71px;

.asset-search {
  margin-top: $basic-spacing;
  &-input {
    margin-bottom: $basic-spacing;
  }
  &-list {
    height: calc(#{$asset-item-height} * 5);
    overflow-y: auto;
    margin-left: -#{$basic-spacing_big};
    margin-right: -#{$basic-spacing_big};
    &_info,
    &_empty {
      @include hint-text;
      padding-left: $basic-spacing_big;
      padding-right: $basic-spacing_big;
    }
    &_empty {
      text-align: center;
    }
    .asset {
      align-items: center;
      height: $asset-item-height;
      padding: 0 $basic-spacing_big;
      &:hover, &.selected {
        background-color: var(--s-color-base-background-hover);
        cursor: pointer;
      }
      &-logo {
        margin-right: $basic-spacing;
        @include asset-logo-styles(40px);
      }
      &-description {
        flex: 1;
        flex-direction: column;
        line-height: $line-height_medium;
        &_symbol {
          font-feature-settings: var(--s-font-feature-settings-common);
          font-weight: 600;
        }
        &_info {
          @include hint-text;
          .asset-id {
            outline: none;
            &:hover {
              text-decoration: underline;
              cursor: pointer;
            }
          }
        }
      }
    }
  }
  &-actions {
    > * {
      flex: .5;
    }
  }
}
</style>
