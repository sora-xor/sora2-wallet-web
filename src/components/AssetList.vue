<template>
  <s-scrollbar class="asset-list" :style="style">
    <div v-if="empty" class="asset-list-empty">
      <slot name="empty" />
    </div>

    <template v-for="(asset, index) in assets">
      <slot v-bind="{ asset, index }" />

      <s-divider v-if="divider && index !== assets.length - 1" :key="`${index}-divider`" />
    </template>
  </s-scrollbar>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator';

import type { Asset } from '@sora-substrate/util';

@Component
export default class AssetList extends Mixins() {
  @Prop({ default: () => [], type: Array }) readonly assets!: Array<Asset>;
  @Prop({ default: 5, type: Number }) readonly items!: number;
  @Prop({ default: false, type: Boolean }) readonly divider!: boolean;

  get empty(): boolean {
    return this.assets.length === 0;
  }

  get style() {
    const dividersHeight = this.divider ? this.items - 1 : 0;
    const height = `calc(var(--s-asset-item-height) * ${this.items} + ${dividersHeight}px)`;

    return {
      height,
    };
  }
}
</script>

<style lang="scss">
.asset-list {
  @include scrollbar($basic-spacing-big);

  &-empty {
    margin-top: $basic-spacing-medium;
    text-align: center;
    @include hint-text;
  }

  .el-divider {
    margin: 0;
  }
}
</style>
