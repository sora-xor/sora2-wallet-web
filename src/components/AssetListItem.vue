<template>
  <div :class="['s-flex', 'asset', { 'asset--with-fiat': withFiat }]" v-bind="$attrs" v-on="$listeners">
    <div class="asset-logo" :class="iconClasses" :style="iconStyles" @click="handleOpenAssetDetails" />
    <nft-token-logo :asset="asset" class="asset-logo__nft-image" />
    <div class="asset-description s-flex">
      <slot name="value" v-bind="asset">
        <div class="asset-symbol">{{ asset.symbol }}</div>
      </slot>
      <token-address :name="asset.name" :symbol="asset.symbol" :address="asset.address" class="asset-info" />
      <slot name="append" v-bind="asset" />
    </div>
    <slot v-bind="asset" />
  </div>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator';
import { Action } from 'vuex-class';

import NftTokenLogo from './NftTokenLogo.vue';

import TranslationMixin from './mixins/TranslationMixin';
import TokenAddress from './TokenAddress.vue';

import { RouteNames } from '../consts';
import { getAssetIconStyles, getAssetIconClasses } from '../util';

import type { Asset } from '@sora-substrate/util/build/assets/types';

@Component({
  components: {
    NftTokenLogo,
    TokenAddress,
  },
})
export default class AssetListItem extends Mixins(TranslationMixin) {
  @Prop({ required: true, type: Object }) readonly asset!: Asset;
  @Prop({ default: false, type: Boolean }) readonly accountAsset!: boolean;
  @Prop({ default: false, type: Boolean }) readonly withFiat!: boolean;

  @Action navigate!: (options: { name: string; params?: object }) => Promise<void>;

  get iconStyles(): object {
    return getAssetIconStyles(this.asset.address);
  }

  get iconClasses(): Array<string> {
    return getAssetIconClasses(this.asset);
  }

  handleOpenAssetDetails(event: Event): void {
    if (this.accountAsset) {
      event.stopImmediatePropagation();
      this.$emit('show-details', this.asset);
      this.navigate({ name: RouteNames.WalletAssetDetails, params: { asset: this.asset } });
    }
  }
}
</script>

<style lang="scss" scoped>
@import '../styles/icons';

.asset {
  align-items: center;
  height: var(--s-asset-item-height);

  &--with-fiat {
    height: var(--s-asset-item-height--fiat);
  }

  &-logo {
    flex-shrink: 0;
    @include asset-logo-styles(42px);

    &:hover {
      cursor: pointer;
    }

    &__nft-image {
      border-radius: 50%;
      object-fit: cover;
      width: var(--s-size-medium);
      height: var(--s-size-medium);
      position: absolute;
      background-color: var(--s-color-base-background);
    }
  }

  &-description {
    flex: 1;
    flex-direction: column;
    align-items: flex-start;
    line-height: var(--s-line-height-big);
    padding: 0 $basic-spacing-small;
    width: 30%;
  }

  &-symbol {
    font-size: var(--s-font-size-big);
    font-weight: 600;
    letter-spacing: var(--s-letter-spacing-small);
    line-height: var(--s-line-height-extra-small);
  }
}
</style>
