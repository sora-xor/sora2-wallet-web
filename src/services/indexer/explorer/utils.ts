import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
