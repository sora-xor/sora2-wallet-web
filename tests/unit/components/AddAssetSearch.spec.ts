import { useDescribe, useShallowMount, useVuex } from '../../utils';
import { MOCK_ACCOUNT_ASSETS, MOCK_ASSETS, MOCK_WHITE_LIST } from '../../utils/mock';
import { MOCK_ADD_ASSET_SEARCH } from '../../utils/AddAssetSearchMock';

import AddAssetSearch from '@/components/AddAsset/AddAssetTokenTab.vue';

import AssetList from '@/components/AssetList.vue';
import AssetListItem from '@/components/AssetListItem.vue';

const accountAssetsAddressTableMock = MOCK_ACCOUNT_ASSETS.reduce((param, item) => {
  param[item.address] = item;
  return param;
}, {});

const createStore = () =>
  useVuex({
    account: {
      state: () => ({
        assets: MOCK_ASSETS,
        accountAssets: MOCK_ACCOUNT_ASSETS,
      }),
      getters: {
        accountAssetsAddressTable: () => accountAssetsAddressTableMock,
        whitelist: () => MOCK_WHITE_LIST,
      },
    },
    router: {
      mutations: {
        navigate: jest.fn(),
      },
    },
  });

useDescribe('AddAssetSearch.vue', AddAssetSearch, () => {
  MOCK_ADD_ASSET_SEARCH.map((item) =>
    it(`[${item.title}]: should be rendered correctly`, () => {
      const wrapper = useShallowMount(AddAssetSearch, {
        store: createStore(),
        data: () => {
          return {
            loading: item.loading,
            search: item.search,
          };
        },
        stubs: {
          AssetList,
          AssetListItem,
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    })
  );
});
