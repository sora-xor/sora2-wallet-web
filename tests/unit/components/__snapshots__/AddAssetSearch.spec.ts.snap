// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAssetSearch.vue should renders correctly 1`] = `
<div
  class="asset-search"
>
  <s-input-stub
    accept=""
    autocomplete="off"
    borderradius="mini"
    class="asset-search-input"
    form=""
    label=""
    maxlength="100"
    name=""
    placeholder=""
    prefix=""
    size="big"
    suffix=""
    tabindex=""
    type="text"
    value=""
  />
   
  <div
    class="asset-search-list"
  >
    <div
      class="asset-search-list_info"
    >
      
      
    
    </div>
     
    <!---->
     
  </div>
   
  <div
    class="asset-search-actions s-flex"
  >
    <s-button-stub
      borderradius="small"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="secondary"
    >
      
      
    
    </s-button-stub>
     
    <s-button-stub
      borderradius="small"
      disabled="true"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="primary"
    >
      
      
    
    </s-button-stub>
  </div>
</div>
`;
